stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -14,434 +14,240 @@
 
 namespace stan {
 namespace math {
+namespace internal {
 
-template <typename T>
-struct value_type {
-  using type = double;
-};
-
-template <>
-struct value_type<int> {
-  using type = int;
-};
-
-template <template <typename> class T_struct, typename T_value>
-struct value_type<T_struct<T_value>> {
-  using type = T_struct<typename value_type<T_value>::type>;
-};
-
-template <typename T_value>
-struct value_type<std::vector<T_value>> {
-  using type = std::vector<typename value_type<T_value>::type>;
-};
+template <typename ReduceFunction, typename ReturnType, typename M,
+          typename... Args>
+struct reduce_sum_impl<ReduceFunction, require_var_t<ReturnType>, ReturnType, M,
+                       Args...> {
+  // TODO(Steve): Move this somewhere smarter
+  // Fails to compile if type T does not have member operator(Integral)
+  template <typename T>
+  using operator_paren_access_t = decltype(std::declval<T>()(int{}));
 
-template <typename T_value, int R, int C>
-struct value_type<Eigen::Matrix<T_value, R, C>> {
-  using type = Eigen::Matrix<typename value_type<T_value>::type, R, C>;
-};
+  struct recursive_reducer {
+    size_t num_terms_;
+    const std::vector<M>& vmapped_;
```

Comment:
If we accepted an eigen matrix here would we want the operations to happen by row?

Reply:
I honestly wouldn't mind putting a scalar_seq_view here, but probably just leaving it as ```real[]``` is the easiest and makes the most sense.

Reply:
`scalar_seq_view` is not a bad idea, it would be nice if users could just pass in a `vector` as well

A matrix would be nice as well, then it's a question of if the function operates elementwise, row, col, etc.

Reply:
Yes, this is what we should allow for: Let users give as slicing argument an array of whatever. So a vector of matrices, Eigen-vectors, Eigen-row-vectors, vectors , etc.

Whatever you slice over will be handled in the most efficient way as compared to the shared arguments.

(but this is can be a follow-up feature)

Reply:
I don't think it would be much elbow grease to allow for `real[]` and `vector[]`. Where we have `vector<M>` now we just use M as the member and in the templates add `typename Var, require_vector_like_t<Var>...`

Reply:
> Let users give as slicing argument an array of whatever.

Oh, hadn't thought of that at all. I like that.